### PR TITLE
add avatar caching #85

### PR DESCRIPTION
**Pull request by anatolym**
Reduced avatar caching TTL to 7 days and fixed ETag generation.

### CODE REVIEW

File: app/rest/proxy/avatar_store_test.go [hunk 1]
```diff
@@ -55,9 +56,18 @@ func TestAvatarStore_Get(t *testing.T) {
...
-	err := ioutil.WriteFile("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", []byte("something"), 0666)
+
+	// file not exists
+	r, size, err := p.Get("some_random_name.image")
+	// nil, 0, errors.Wrapf(err, "can't load avatar %s, id")
+	assert.Nil(t, r)
+	assert.Equal(t, 0, size)
+	assert.EqualError(t, err, "can't load avatar some_random_name.image, id: open /tmp/avatars.test/91/some_random_name.image: no such file or directory")
+	// file exists
+	err = ioutil.WriteFile("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", []byte("something"), 0666)
```
Comment (golangci) [inline app/rest/proxy/avatar_store_test.go]: G303: File creation in shared tmp directory without using ioutil.Tempfile

File: app/rest/proxy/avatar_store_test.go [hunk 2]
```diff
@@ -133,3 +143,31 @@ func TestAvatarStore_resize(t *testing.T) {
...
+func TestAvatarStore_ID(t *testing.T) {
+	p := NewFSAvatarStore("/tmp/avatars.test", 300)
+	os.MkdirAll("/tmp/avatars.test/30", 0700)
+	defer os.RemoveAll("/tmp/avatars.test")
+
+	// file not exists
+	id := p.ID("some_random_name.image")
+	assert.Equal(t, "a008de0a2ccb3308b5d99ffff66436e15538f701", id) // store.EncodeID("some_random_name.image")
+	// file exists
+	ioutil.WriteFile("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", []byte("something"), 0666)
```
Comment (golangci) [inline app/rest/proxy/avatar_store_test.go]: G303: File creation in shared tmp directory without using ioutil.Tempfile

File: app/rest/proxy/avatar_store_test.go [hunk 3]
```diff
@@ -133,3 +143,31 @@ func TestAvatarStore_resize(t *testing.T) {
...
+	err := os.Chtimes("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", touch, touch)
+	require.NoError(t, err)
+	id = p.ID("b3daa77b4c04a9551b8781d03191fe098f325e67.image")
+	assert.Equal(t, "325d5b451f32c2f8e7f30a9fd65bff6a42954d9a", id) // store.EncodeID("b3daa77b4c04a9551b8781d03191fe098f325e67.image1500000000")
+}
+func BenchmarkAvatarStore_ID(b *testing.B) {
+	p := NewFSAvatarStore("/tmp/avatars.test", 300)
+	os.MkdirAll("/tmp/avatars.test/30", 0700)
+	defer os.RemoveAll("/tmp/avatars.test")
+	ioutil.WriteFile("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", []byte("something"), 0666)
```
Comment (golangci) [inline app/rest/proxy/avatar_store_test.go]: G303: File creation in shared tmp directory without using ioutil.Tempfile

File: app/rest/proxy/avatar_store_test.go [hunk 4]
```diff
@@ -133,3 +143,31 @@ func TestAvatarStore_resize(t *testing.T) {
 		assert.Equalf(t, c.hr, bounds.Dy(), "file %s", c.file)
 	}
 }
+
+func TestAvatarStore_ID(t *testing.T) {
+	p := NewFSAvatarStore("/tmp/avatars.test", 300)
+	os.MkdirAll("/tmp/avatars.test/30", 0700)
```
Comment (golangci) [inline app/rest/proxy/avatar_store_test.go]: Error return value of `os.MkdirAll` is not checked

File: app/rest/proxy/avatar_store_test.go [hunk 5]
```diff
@@ -133,3 +143,31 @@ func TestAvatarStore_resize(t *testing.T) {
...
+	ioutil.WriteFile("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", []byte("something"), 0666)
+	touch := time.Date(2017, 7, 14, 2, 40, 0, 0, time.UTC) // 1500000000
+	err := os.Chtimes("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", touch, touch)
+	require.NoError(t, err)
+	id = p.ID("b3daa77b4c04a9551b8781d03191fe098f325e67.image")
+	assert.Equal(t, "325d5b451f32c2f8e7f30a9fd65bff6a42954d9a", id) // store.EncodeID("b3daa77b4c04a9551b8781d03191fe098f325e67.image1500000000")
+}
+func BenchmarkAvatarStore_ID(b *testing.B) {
+	p := NewFSAvatarStore("/tmp/avatars.test", 300)
+	os.MkdirAll("/tmp/avatars.test/30", 0700)
```
Comment (golangci) [inline app/rest/proxy/avatar_store_test.go]: Error return value of `os.MkdirAll` is not checked

File: app/rest/proxy/avatar_store_test.go [hunk 6]
```diff
@@ -133,3 +145,34 @@ func TestAvatarStore_resize(t *testing.T) {
...
+	p := NewFSAvatarStore("/tmp/avatars.test", 300)
+	err := os.MkdirAll("/tmp/avatars.test/30", 0700)
+	require.NoError(t, err)
+	defer os.RemoveAll("/tmp/avatars.test")
+
+	// file not exists
+	id := p.ID("some_random_name.image")
+	assert.Equal(t, "a008de0a2ccb3308b5d99ffff66436e15538f701", id) // store.EncodeID("some_random_name.image")
+	// file exists
+	err = ioutil.WriteFile("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", []byte("something"), 0666)
```
Comment (golangci) [inline app/rest/proxy/avatar_store_test.go]: G303: File creation in shared tmp directory without using ioutil.Tempfile

File: app/rest/proxy/avatar_store_test.go [hunk 7]
```diff
@@ -133,3 +145,34 @@ func TestAvatarStore_resize(t *testing.T) {
...
+	err = os.Chtimes("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", touch, touch)
+	require.NoError(t, err)
+	id = p.ID("b3daa77b4c04a9551b8781d03191fe098f325e67.image")
+	assert.Equal(t, "325d5b451f32c2f8e7f30a9fd65bff6a42954d9a", id) // store.EncodeID("b3daa77b4c04a9551b8781d03191fe098f325e67.image1500000000")
+}
+func BenchmarkAvatarStore_ID(b *testing.B) {
+	p := NewFSAvatarStore("/tmp/avatars.test", 300)
+	os.MkdirAll("/tmp/avatars.test/30", 0700)
+	defer os.RemoveAll("/tmp/avatars.test")
+	err := ioutil.WriteFile("/tmp/avatars.test/30/b3daa77b4c04a9551b8781d03191fe098f325e67.image", []byte("something"), 0666)
```
Comment (golangci) [inline app/rest/proxy/avatar_store_test.go]: G303: File creation in shared tmp directory without using ioutil.Tempfile

File: app/rest/proxy/avatar_store_test.go (hunk 6 above)
Comment (anatolym) [reply app/rest/proxy/avatar_store_test.go]: @umputun should I do something with this warns?

Comment (umputun) [reply app/rest/proxy/avatar_store_test.go]: nah, this is bs, especially for tests.

**Pull request merged**